segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -293,17 +295,54 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
                         log.debug("{}: deleteGarbage - Segment is not marked as deleted. segment={}.", traceObjectId, streamSegmentName);
                         return CompletableFuture.completedFuture(null);
                     } else {
-                        return new ChunkIterator(storageExecutor, txn, segmentMetadata)
-                                .forEach((metadata, name) -> {
-                                    metadata.setActive(false);
-                                    txn.update(metadata);
-                                    chunksToDelete.add(name);
-                                })
+                        val chunksToDelete = Collections.synchronizedSet(new HashSet<String>());
+                        val currentBatch = Collections.synchronizedSet(new HashSet<ChunkMetadata>());
+                        val currentChunkName = new AtomicReference<String>(segmentMetadata.getFirstChunk());
+
+                        return Futures.loop(
+                                () -> null != currentChunkName.get(),
+                                () -> txn.get(currentChunkName.get())
+                                        .thenComposeAsync(metadata -> {
+                                            val chunkMetadata = (ChunkMetadata) metadata;
+                                            CompletableFuture<Void> retFuture = CompletableFuture.completedFuture(null);
+
+                                            // Add to list of chunks to delete
+                                            chunksToDelete.add(chunkMetadata.getName());
+
+                                            // Skip if metadata is possibly deleted in last attempt, we are done.
+                                            if (null == chunkMetadata) {
```

Comment:
How can this be null? The previous line assumes it's not.

Reply:
fixed